[JSC] Optimize and clean up Min / Max

Constellation opened the pull request:
#### 81e9f6868ad1953daa05923a488fc205322fceb3
<pre>
[JSC] Optimize and clean up Min / Max
<a href="https://bugs.webkit.org/show_bug.cgi?id=276694">https://bugs.webkit.org/show_bug.cgi?id=276694</a>
<a href="https://rdar.apple.com/131882633">rdar://131882633</a>

Reviewed by Justin Michaud.

This patch does holistic Math.min / Math.max optimization.

1. We add thunks for them in ARM64, which have super fast path for Int32-Int32 / Int32-Double / Double-Int32 / Double-Double. (We do not use it for now, we need a bit tweaking)
2. Use ARM64 fmax / fmin in DFG and FTL JIT and thunk. The semantics of fmax / fmin is aligned to JS&apos;s semantics, so we can leverage it.
3. Add jsMaxDouble / jsMinDouble and use them in Math.max / Math.min runtime functions.
4. Use inline asm to use fmax / fmin in jsMaxDouble / jsMinDouble in ARM64.

Simple FTL perf is getting 11% faster.

                               ToT                     Patched

    math-max             10.4292+-0.1181     ^      9.3465+-0.0937        ^ definitely 1.1158x faster

And if we disable DFG / FTL, Math.min / Math.max are 1.9x faster.

                               ToT                     Patched

    math-max             50.0668+-0.3260     ^     26.2990+-0.0774        ^ definitely 1.9038x faster

* JSTests/microbenchmarks/math-max.js: Added.
(max):
* Source/JavaScriptCore/b3/B3Common.h:
(JSC::B3::fMax): Deleted.
(JSC::B3::fMin): Deleted.
* Source/JavaScriptCore/b3/B3ConstDoubleValue.cpp:
(JSC::B3::ConstDoubleValue::fMinConstant const):
(JSC::B3::ConstDoubleValue::fMaxConstant const):
* Source/JavaScriptCore/b3/B3ConstFloatValue.cpp:
(JSC::B3::ConstFloatValue::fMinConstant const):
(JSC::B3::ConstFloatValue::fMaxConstant const):
* Source/JavaScriptCore/dfg/DFGOperations.cpp:
(JSC::DFG::JSC_DEFINE_NOEXCEPT_JIT_OPERATION):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileArithMinOrMax):
* Source/JavaScriptCore/ftl/FTLOutput.cpp:
(JSC::FTL::Output::doubleMax):
(JSC::FTL::Output::doubleMin):
* Source/JavaScriptCore/ftl/FTLOutput.h:
* Source/JavaScriptCore/jit/ThunkGenerators.cpp:
(JSC::maxThunkGenerator):
(JSC::minThunkGenerator):
* Source/JavaScriptCore/jit/ThunkGenerators.h:
* Source/JavaScriptCore/runtime/MathCommon.h:
(JSC::Math::fMax):
(JSC::Math::fMin):
(JSC::Math::jsMaxDouble):
(JSC::Math::jsMinDouble):
* Source/JavaScriptCore/runtime/MathObject.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/VM.cpp:
(JSC::thunkGeneratorForIntrinsic):

Canonical link: <a href="https://commits.webkit.org/281118@main">https://commits.webkit.org/281118@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0a536641b39d914cd3b126116daeffe5237e989d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/58826 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/38154 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/11316 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/62458 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/9270 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/45791 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/9470 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/47581 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wincairo-tests~~](https://ews-build.webkit.org/#/builders/60/builds/6601 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/60857 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/35699 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/50870 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/28437 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/32431 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/8160 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/8274 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/51919 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/54387 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/8440 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/64159 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/58068 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/2739 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/8421 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/54904 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/2748 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/50896 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/54998 "Passed tests") | 
| | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/88/builds/2296 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/79829 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8774 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/33984 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/13808 "Passed tests") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/35068 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/36153 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/34814 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

justinmichaud reviewed (approved):
r=me. I am assuming we have tests for all the weird NaN cases.

justinmichaud commented on Source/JavaScriptCore/b3/B3ConstDoubleValue.cpp:
```diff
@@ -140,14 +140,14 @@ Value* ConstDoubleValue::fMinConstant(Procedure& proc, const Value* other) const
     if (!other->hasDouble())
         return nullptr;
-    return proc.add<ConstDoubleValue>(origin(), fMin(m_value, other->asDouble()));
+    return proc.add<ConstDoubleValue>(origin(), Math::fMin(m_value, other->asDouble()));
```
This can be used from wasm, which can observe non-canonical NaNs, right?

Constellation replied:
This is aligned to fmax behavior to ARM64, and no change is done in this patch.

Constellation committed:
[JSC] Optimize and clean up Min / Max
https://bugs.webkit.org/show_bug.cgi?id=276694
rdar://131882633

Reviewed by Justin Michaud.

This patch does holistic Math.min / Math.max optimization.

1. We add thunks for them in ARM64, which have super fast path for Int32-Int32 / Int32-Double / Double-Int32 / Double-Double. (We do not use it for now, we need a bit tweaking)
2. Use ARM64 fmax / fmin in DFG and FTL JIT and thunk. The semantics of fmax / fmin is aligned to JS's semantics, so we can leverage it.
3. Add jsMaxDouble / jsMinDouble and use them in Math.max / Math.min runtime functions.
4. Use inline asm to use fmax / fmin in jsMaxDouble / jsMinDouble in ARM64.

Simple FTL perf is getting 11% faster.

                               ToT                     Patched

    math-max             10.4292+-0.1181     ^      9.3465+-0.0937        ^ definitely 1.1158x faster

And if we disable DFG / FTL, Math.min / Math.max are 1.9x faster.

                               ToT                     Patched

    math-max             50.0668+-0.3260     ^     26.2990+-0.0774        ^ definitely 1.9038x faster

* JSTests/microbenchmarks/math-max.js: Added.
(max):
* Source/JavaScriptCore/b3/B3Common.h:
(JSC::B3::fMax): Deleted.
(JSC::B3::fMin): Deleted.
* Source/JavaScriptCore/b3/B3ConstDoubleValue.cpp:
(JSC::B3::ConstDoubleValue::fMinConstant const):
(JSC::B3::ConstDoubleValue::fMaxConstant const):
* Source/JavaScriptCore/b3/B3ConstFloatValue.cpp:
(JSC::B3::ConstFloatValue::fMinConstant const):
(JSC::B3::ConstFloatValue::fMaxConstant const):
* Source/JavaScriptCore/dfg/DFGOperations.cpp:
(JSC::DFG::JSC_DEFINE_NOEXCEPT_JIT_OPERATION):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileArithMinOrMax):
* Source/JavaScriptCore/ftl/FTLOutput.cpp:
(JSC::FTL::Output::doubleMax):
(JSC::FTL::Output::doubleMin):
* Source/JavaScriptCore/ftl/FTLOutput.h:
* Source/JavaScriptCore/jit/ThunkGenerators.cpp:
(JSC::maxThunkGenerator):
(JSC::minThunkGenerator):
* Source/JavaScriptCore/jit/ThunkGenerators.h:
* Source/JavaScriptCore/runtime/MathCommon.h:
(JSC::Math::fMax):
(JSC::Math::fMin):
(JSC::Math::jsMaxDouble):
(JSC::Math::jsMinDouble):
* Source/JavaScriptCore/runtime/MathObject.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/VM.cpp:
(JSC::thunkGeneratorForIntrinsic):

Canonical link: https://commits.webkit.org/281118@main
